src/box/memtx_tree.cc

```diff
@@ -965,23 +1042,60 @@ tree_iterator_start(struct iterator *iterator, struct tuple **ret)
 	}
 
 /********MVCC TRANSACTION MANAGER STORY GARBAGE COLLECTION BOUND START*********/
+	if (it->skip != 0) {
+		if (res == NULL || !eq_match) {
+			/*
+			 * We have stepped over some amount of tuples and got to
+			 * the end of the index or stepped over the matching set
+			 * (if iterator is EQ or REQ). Lets inform MVCC like we
+			 * have counted tuples in the index by our iterator and
+			 * key. Insertion or deletion of any matching tuple into
+			 * the index will conflict with us.
+			 */
+			memtx_tx_track_count(txn, space, index_base,
+					     type, start_data.key,
+					     start_data.part_count);
+
+		} else {
+			/*
+			 * We have stepped over some amount of tuples and got to
+			 * a tuple. Changing the amount of matching tuples prior
+			 * to the approached one must conflict with us, so lets
+			 * inform MVCC like we have counted tuples in the index
+			 * by our key and iterator until the approached tuple.
+			 *
+			 * The approached tuple itself is read above, so its
+			 * replacement or deletion is tracked already.
+			 */
+			memtx_tx_track_count_until(txn, space, index_base,
+						   type, start_data.key,
+						   start_data.part_count,
+						   res->tuple, res->hint);
+		}
+	}
 	/*
 	 * If the key is full then all parts present, so EQ and REQ iterators
 	 * can return no more than one tuple.
 	 */
 	struct key_def *cmp_def = index->base.def->cmp_def;
 	bool key_is_full = start_data.part_count == cmp_def->part_count;
-	if (key_is_full && !eq_match)
+	/*
+	 * We track all the skipped tuples using one of count trackers above,
+	 * no extra tracking is required in this case, insertion or deletion
+	 * of a matching tuple will be caught by the count tracker.
+	 */
+	if (it->skip == 0 && key_is_full && !eq_match)
 		memtx_tx_track_point(txn, space, index_base, it->key_data.key);
 	/*
 	 * Since MVCC operates with `key_def` of index but `start_data` can
 	 * contain key extracted with `cmp_def`, we should crop it by passing
 	 * `part_count` not greater than `key_def->part_count`.
 	 */
 	uint32_t key_part_count = index->base.def->key_def->part_count;
-	if (!key_is_full ||
-	    ((type == ITER_GE || type == ITER_LE) && !equals) ||
-	    (type == ITER_GT || type == ITER_LT))
+	if (it->skip == 0 &&
```

Comment:
Nit: maybe add a `goto` instead of `it->skip == 0`?

Reply:
Done.